main.js

```diff
@@ -282,44 +314,57 @@ loops = {
 	},
 
 	upgrade: () => {
+		// leave if all firewalls are upgraded to max
+		if (!vars.fireWall[3].needUpgrade)
+			return;
+		const i = getRandomInt(1,3);
+		// if this fireWall is already fully upgraded, get an other random firewall.
+		if (!vars.fireWall[i - 1].needUpgrade)
+			vars.loops.upgrade();
 		vars.balance = parseInt($("#window-my-coinamount").text());
-		// if the back button is visible, we're on a page, let's back out
+		// if the back button is visible, we're on a page, let's back out and hide the firewall warning.
 		if ($("#window-firewall-pagebutton").is(":visible") === true) {
+			$("#tutorial-firewall").css("display","none");
 			$("#window-firewall-pagebutton").click();
 		}
+
 		// just get a random port, because who cares
-		const firewall = getRandomInt(1,3);
 		// select the firewall
-		log(`. Handling upgrades to firewall ${firewall}`);
-		$(`#window-firewall-part${firewall}`).click();
+		log(`. Handling upgrades to firewall ${vars.fireWall[i].name}`);
+		$(`#window-firewall-part${i}`).click();
 		// get stats
-		const stats = {
-			charge: parseInt($("#shop-max-charges").text()),
-			strength: parseInt($("#shop-strength").text()),
-			regen: parseInt($("#shop-regen").text())
-		};
-		const statLookup = {
-			charge: "max_charge10",
-			strength: "difficulty",
-			regen: "regen"
-		};
-		const maxStats = {
-			charge: 30,
-			strength: 4,
-			regen: 10
-		};
-
-		for (const stat in stats) {
+		const stats = [
+			parseInt($("#shop-max-charges").text()), parseInt($("#shop-strength").text()), parseInt($("#shop-regen").text())
+		];
+		const statLookup = [
+			"max_charge10", "difficulty", "regen"
+		];
+		const maxStats = [
+			30, 4, 10
+		];
+		let maxUpgradeCount = 0;
+		for (const stat in maxStats) {
 			if (stats[stat] < maxStats[stat]) {
 				const statPrice = parseInt($(`#shop-firewall-${statLookup[stat]}-value`).text());
-				if (statPrice < vars.balance) {
-					log(`. Buying ${stat}`);
+				if (statPrice < (vars.balance / 3)) {
```

Comment:
Why are we dividing by 3 here?  I assume it's to prevent upgrading from taking all our money and us not being able to hack?  If that's the case, I'm fine with it, but I'd like to keep magic numbers out of the code.  Maybe move it to config and call it `maxUpgradeCost`, and change it to a percentage (`.33`) and label it accordingly?